### PR TITLE
Opensea NFT Trade classifier

### PR DESCRIPTION
**Pull request by sketsdever**
Creates a new classifier for NFT trades. A trade stores:
- Buyer/seller
- Price (minus exchange fees)
- Collection address and token id

Example block: [13796299](https://etherscan.io/block/13796299)

Edits welcome - I expect we'll adjust / build on top of this.

### CODE REVIEW

File: alembic/versions/3c54832385e3_create_nft_trades_table.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        sa.Column("abi_name", sa.String(1024), nullable=False),
+        sa.Column("transaction_hash", sa.String(66), nullable=False),
+        sa.Column("transaction_position", sa.Numeric, nullable=False),
+        sa.Column("block_number", sa.Numeric, nullable=False),
+        sa.Column("trace_address", sa.String(256), nullable=False),
+        sa.Column("protocol", sa.String(256), nullable=False),
+        sa.Column("error", sa.String(256), nullable=True),
+        sa.Column("seller_address", sa.String(256), nullable=False),
+        sa.Column("buyer_address", sa.String(256), nullable=False),
+        sa.Column("payment_token", sa.String(256), nullable=False),
```
Comment (lukevs) [inline alembic/versions/3c54832385e3_create_nft_trades_table.py:31]: `payment_token_address` for consistency 🙏

File: mev_inspect/classifiers/specs/opensea.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class OpenseaClassifier(NftTradeClassifier):
+    @staticmethod
+    def parse_trade(
+        trace: DecodedCallTrace,
+        child_transfers: List[Transfer],
+    ) -> Optional[NftTrade]:
+        addresses = trace.inputs.get("addrs")
+        if addresses is None:
```
Comment (lukevs) [inline mev_inspect/classifiers/specs/opensea.py:19]: When is this missing?

File: mev_inspect/classifiers/helpers.py [hunk 3]
```diff
@@ -1,10 +1,67 @@
...
 from mev_inspect.schemas.transfers import ETH_TOKEN_ADDRESS, Transfer
 
 
+def create_nft_trade_from_transfers(
+    trace: DecodedCallTrace,
+    child_transfers: List[Transfer],
+    collection_address: str,
+    seller_address: str,
+    buyer_address: str,
+    exchange_wallet_address: Optional[str],
```
Comment (lukevs) [inline mev_inspect/classifiers/helpers.py:15]: looks like at least for now this is never None

File: mev_inspect/classifiers/specs/opensea.py (hunk 2 above)
Comment (sketsdever) [reply mev_inspect/classifiers/specs/opensea.py:19]: Theoretically never. But the linter throws an error here if you don't check the None case - let me know if there's a better way around this

Comment (sketsdever) [reply mev_inspect/classifiers/specs/opensea.py:19]: Theoretically never. But the linter throws an error if you don't check the None case - is there a better way to handle?

Comment (lukevs) [reply mev_inspect/classifiers/specs/opensea.py:19]: If you get the key explicitly it'll avoid the none type and just throw a KeyError if it's missing
```
addresses = trace.inputs["addrs"]
```

File: mev_inspect/inspect_block.py [hunk 4]
```diff
@@ -209,6 +219,11 @@ async def inspect_many_blocks(
     )
     write_punk_snipes(inspect_db_session, punk_snipes)
 
+    delete_nft_trades_for_blocks(
+        inspect_db_session, after_block_number, before_block_number
+    )
+    write_nft_trades(inspect_db_session, nft_trades)
```
Comment (lukevs) [inline mev_inspect/inspect_block.py:225]: a bug (that exists in main as well, PR coming momentarily) is this should be `all_nft_trades)

File: alembic/versions/3c54832385e3_create_nft_trades_table.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+Revises: 15ba9c27ee8a
+Create Date: 2021-12-19 22:50:28.936516
+
+"""
+import sqlalchemy as sa
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision = "3c54832385e3"
+down_revision = "15ba9c27ee8a"
```
Comment (lukevs) [inline alembic/versions/3c54832385e3_create_nft_trades_table.py:13]: latest on main is `4b9d289f2d74`

**Review by lukevs — APPROVED** [review]

**Pull request merged**